doc/developer/design/20240508_temporal_filter_expressions.md

```diff
@@ -0,0 +1,62 @@
+
+## Success Criteria
+
+Users are able to express at least simple operations on `mz_now()` in temporal filters (the above `WHERE` should function).
+
+## Out of Scope
+
+The [braindump](https://www.notion.so/materialize/mz_now-braindump-3acde85e02d84cb09808d7c63dab009a#3a56d5b83d554ccfb630300b02ffe5aa) for this feature suggested
+
+> For the mz_timestamp type, ... allow the same operators as are allowed for the timestamp type.
+
+We can perhaps solve the immediate problem only implementing a small subset of those operators.
+
+## Solution Proposal
+
+Three separate features can be implemented:
+
+1. Allow more types to be implicitly castable to `mz_timestamp`.
+Most time-adjacent types already support this.
+We should implement the rest (`time`, `date`).
+
+1. Allow `mz_timestamp` types to be added and subtracted to other `mz_timestamp`s (overflows will error).
+Many types are [castable to `mz_timestamp`](https://materialize.com/docs/sql/types/mz_timestamp/) implicitly.
+If we allow the addition and subtraction of `mz_timestamp` to `mz_timestamp`,
+then expressions like `mz_now() - '30s'::interval` will implicitly cast the `'30s'::interval` to an `mz_timestamp`,
+and the subtraction from `mz_now()` will work.
+This approach has the benefit that we never need to worry about casting `mz_timestamp` to another type (like `timestamp`) that has different precision, range, and semantics.
+
+1. Teach temporal filters in `linear.rs` to recognize expressions like `mz_now() + some_mz_timestamp > x` and convert them to `mz_now() > x - some_mz_timestamp`.
+(`-` and the other inequality operators also supported.)
+This logic can happen repeatedly, so `mz_now() + x - y <= z` becomes `mz_now() <= z - x + y`.
+
+A concern about this is that the underlying data gain additional restrictions in order to not produce errors.
+For example, `mz_now() - '10 years'::interval > x` is converted into `mz_now() > x + '10 years'::interval`,
+so if `x` is within 10 years of u64::MAX milliseconds, the entire materialized view will error and be unreadable at any time.
+Due to the range of `mz_timestamp` this seems unlikely to ever matter.
+We already have a tiny version of this problem too: the `=`, `<=`, and `>` inequalities all add 1 to the non-`mz_now()` side,
+so if the underlying data contain `u64::MAX` and one of those inequalities, the materialized view will error.
+Because we already have a small version of this problem and the `mz_timestamp` range is so large (~500M years),
+allowing the users to reduce that range seems acceptable.
+
+## Minimal Viable Prototype
+
+An MVP [exists](https://github.com/MaterializeInc/materialize/compare/main...maddyblue:materialize:mznow-expr).
+It includes some now-merged code (implicit casting of interval to `mz_timestamp`) since that matches many existing casts and could be safely implemented before this design doc was written.
+
+## Alternatives
+
+
+
+## Open questions
+
+- Is it always safe to move expressions around?
```

Comment:
Nope! :D

At least, if the goal is to do so without altering semantics. I think we don't have a framework in place for resolving that tension: where is it "safe" to modify user queries, and in what framework.

Reply:
I get the feeling from this comment that even if I can come up with a better place to do the expression modification, we may still decide it's not safe to do. Thinking through this with the example:

`WHERE mz_now() + y > x`

this would error when `mz_now() + y` overflows for some positive `y` and underflows for some negative `y`. If we had a rewriter that converted it to

`WHERE mz_now() > x - y`

this would never error based on `mz_now()`'s value, but would instead error based on `x`'s value in relation to `y`. Is this the type of thing you mean by "Nope, at least, if the goal is to do so without altering semantics."?

Reply:
I think the important thing is that we accept that it is semantics altering, and find the right place to elect to do that. I don't think semantics altering transformations are appropriate in the core of the system, but they seem totally fine at higher levels. For example, if there is a session variable `MZ_ERGONOMIC_REWRITES` or something, that elects users in to altering the behavior of their SQL, and accept that e.g.
```sql
CREATE VIEW foo1 AS SELECT * FROM bar;
CREATE MATERIALIZED VIEW foo2 AS SELECT * FROM bar;
SELECT * FROM foo1 EXCEPT ALL SELECT * FROM foo2
```
may produce non-empty results, because `SELECT` can produce different outputs from maintained results, then totally fine. It's imo some amount on Surfaces to figure out where this lands on "surprise and delight", probably a bit of both and not always the good kind.

Reply:
> `WHERE mz_now() + y > x`

I think at the moment this has the property you mentioned, may error may not, but also cannot be materialized. That means that there are not two things that are meant to be the same but are not the same (because we cannot materialize it).

> `WHERE mz_now() > x - y`

This is a thing that errors based on `x` and `y`, and can be materialized, and (should, imo) behave the same in both cases.

The second one is great, in that if they typed `x - y` then there's no surprise should it error, or at least the explanation is right there in front of them. If we converted the first one to the second one automatically, it seems like there are two cases: 
1. We convert *everything*, even `SELECT` statements, and have new behavior that matches between ad-hoc and maintained views, but which doesn't match their original query ("error: x - y overflowed", where "x - y" doesn't appear in their query. The fix to avoiding overflow is to move the subtractions around, but .. we might just undo that for them and error again).
2. We convert only dataflows, leaving `SELECT` statements as they were, and now have a variance between the two implementations of "the same thing". This probably has unpleasant outcomes in corner cases that we eat when we first realize why folks trip over them ("I put a `1-1-1970` in my dataset as a sentinel, but can't read from my index any more because 'subtraction underflow'??").

There could be other cases! I'm pretty opinionated about maintaining consistency between selecting from views and selecting from indexes. I'm less opinionated on how we arrive at what happens when you do that. E.g. a flag that changes the behavior (e.g. MySQL has one that changes from SQL99's `WITH RECURSIVE` semantics to actual fixed point semantics, but you have to opt in).